src/topology/builder.rs

```diff
@@ -343,6 +344,88 @@ pub async fn build_pieces(
 
                 Task::new(key.clone(), typetag, transform)
             }
+            Transform::DispatchFunction(mut t) => {
+                let mut output_fans = IndexMap::new();
+                let mut output_controls = IndexMap::new();
+                let output_names = named_outputs.clone();
+                for name in output_names.iter() {
+                    let (output, control) = Fanout::new();
+                    output_fans.insert(name.clone(), output);
+                    output_controls.insert(name.clone(), control);
+                }
+
+                let mut input_rx = input_rx
+                    .filter(move |event| ready(filter_event_type(event, input_type)))
+                    .ready_chunks(128); // 128 is an arbitrary, smallish constant
+
+                let mut timer = crate::utilization::Timer::new();
+                let mut last_report = Instant::now();
+
+                let transform = async move {
+                    timer.start_wait();
+                    while let Some(events) = input_rx.next().await {
+                        let stopped = timer.stop_wait();
+                        if stopped.duration_since(last_report).as_secs() >= 5 {
+                            timer.report();
+                            last_report = stopped;
+                        }
+
+                        emit!(&EventsReceived {
+                            count: events.len(),
+                            byte_size: events.size_of(),
+                        });
+
+                        // TODO: optimize the creation of buffers
+                        let mut output_bufs = output_names
+                            .iter()
+                            .map(|name| (name.clone(), Vec::with_capacity(events.len())))
+                            .collect::<IndexMap<_, _>>();
+                        let mut bufs = output_names
+                            .iter()
+                            .map(|name| (name.clone(), Vec::with_capacity(1)))
+                            .collect::<IndexMap<_, _>>();
+
+                        for v in events {
+                            t.transform(&mut bufs, v);
```

Comment:
maybe, instead of giving all the possible outputs, modifying it like following

```rust
fn transform(&self, outputs: &mut Vec<(String, Event)>, event: Event)
```

Where, in `outputs` we insert the event for each output only when needed.